hazelcast/src/main/java/com/hazelcast/client/impl/protocol/ClientProtocolErrorCodes.java

```diff
@@ -0,0 +1,237 @@
+            @Override
+            public Throwable createException(String message) {
+                return new AuthenticationException(message);
+            }
+        });
+        register(HAZELCAST_INSTANCE_NOT_ACTIVE, HazelcastInstanceNotActiveException.class, new ExceptionFactory() {
+            @Override
+            public Throwable createException(String message) {
+                return new HazelcastInstanceNotActiveException(message);
+            }
+        });
+        register(ILLEGAL_STATE, IllegalStateException.class, new ExceptionFactory() {
+            @Override
+            public Throwable createException(String message) {
+                return new IllegalStateException(message);
+            }
+        });
+        register(ILLEGAL_ARGUMENT, IllegalArgumentException.class, new ExceptionFactory() {
+            @Override
+            public Throwable createException(String message) {
+                return new IllegalArgumentException(message);
+            }
+        });
+        register(EXECUTION, ExecutionException.class, new ExceptionFactory() {
+            @Override
+            public Throwable createException(String message) {
+                return new ExecutionException(message, null);
+            }
+        });
+        register(TIMEOUT, TimeoutException.class, new ExceptionFactory() {
+            @Override
+            public Throwable createException(String message) {
+                return new TimeoutException(message);
+            }
+        });
+        register(INTERRUPTED, InterruptedException.class, new ExceptionFactory() {
+            @Override
+            public Throwable createException(String message) {
+                return new InterruptedException(message);
+            }
+        });
+        register(ILLEGAL_MONITOR_STATE, IllegalMonitorStateException.class, new ExceptionFactory() {
+            @Override
+            public Throwable createException(String message) {
+                return new IllegalMonitorStateException(message);
+            }
+        });
+        register(DISTRIBUTED_OBJECT_DESTROYED, DistributedObjectDestroyedException.class, new ExceptionFactory() {
+            @Override
+            public Throwable createException(String message) {
+                return new DistributedObjectDestroyedException(message);
+            }
+        });
+        register(ILLEGAL_STATE, IllegalStateException.class, new ExceptionFactory() {
+            @Override
+            public Throwable createException(String message) {
+                return new IllegalStateException(message);
+            }
+        });
+        register(HAZELCAST, HazelcastException.class, new ExceptionFactory() {
+            @Override
+            public Throwable createException(String message) {
+                return new HazelcastException(message);
+            }
+        });
+        //TODO following list will be registered !?!?!
```

Comment:
I would suggest only add the exception if it has a meaning and use for the client implementor and wrap all user unrelated but server related errors in a generic error such as internal error. We should go through each of them and make this analysis. Furthermore, the user will most probably ask us which one is possible error in which operation and what action the user needs to take when a specific error is received (e.g. retry the operation?). What do you think about it?